### PR TITLE
DQN + Atari + JAX

### PR DESCRIPTION
**Pull request by yooceii**
## Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] New algorithm
- [ ] Documentation

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/vwxyzjn/cleanrl/blob/master/CONTRIBUTING.md) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
- [ ] I have updated the tests accordingly (if applicable).

If you are adding new algorithms or your change could result in performance difference, you may need to (re-)run tracked experiments. See https://github.com/vwxyzjn/cleanrl/pull/137 as an example PR. 
- [x] I have contacted [vwxyzjn](https://github.com/vwxyzjn) to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark) (**required**).
- [ ] I have tracked applicable experiments in [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) with `--capture-video` flag toggled on (**required**).
- [ ] I have added additional documentation and previewed the changes via `mkdocs serve`.
    - [ ] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [ ] I have added links to the original paper and related papers (if applicable).
    - [ ] I have added links to the PR related to the algorithm.
    - [x] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [x] I have added the learning curves (in PNG format with `width=500` and `height=300`).
    - [x] I have added links to the tracked experiments.
- [ ] I have updated the tests accordingly (if applicable).

### CODE REVIEW

**Comment by yooceii** [comment]
image

JAX gains roughly 40% SPS increase with same epi_ret performance.

Full report here: https://wandb.ai/yooceii/dqn-atari-jax/reports/DQN-JAX--VmlldzoyMjkzMTM2

**Review by vwxyzjn — COMMENTED** [review]
Thank you @yooceii and @kinalmehta for the awesome independent DQN implementations in this PR and #222, so we can reconcile the differences. 

Things I like about this #231:
1.  it jitted the action sampling which could offer additional speed up, but we don't know for sure.
2. it demonstrated `vmap`, which is pretty cool. That said, my brain is still wired up to deal with batch sizes and feels in this case `vmap` is not bringing too much...

Things I like about #222:
1. Adoption of `TrainState` which gives a cleaner API IMO.

Btw @yooceii I can confirm that `dqn_atari_jax.py` in #222 also gives about 40% or more performance boost (also weird that `dqn_atari` runs slower on RTX 3060 TI than RTX 2060). 

image

All in all, I think the remaining thing is to check if jitting linear scheduling and sampling actions actually improve throughput. If it does, let's do that in #222, otherwise, I feel #222 would be a good candidate to be merged.

**Comment by yooceii** [comment]
image
Looks like combining linear_schedule and select_action and jitting do get a little better performance. Now it gains roughly __50__% more SPS. @vwxyzjn 
Wandb report is also updated.

File: cleanrl/dqn_atari_jax.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+    # TRY NOT TO MODIFY: seeding
+    key = jax.random.PRNGKey(args.seed)
+    q_network_seed, t_network_seed, rng_seed = jax.random.split(key, num=3)
+
+    # env setup
+    envs = gym.vector.SyncVectorEnv([make_env(args.env_id, args.seed, 0, args.capture_video, run_name)])
+    assert isinstance(envs.single_action_space, gym.spaces.Discrete), "only discrete action space is supported"
+    obs = envs.reset()
+    network_def = QNetwork(envs.single_action_space.n, nn.initializers.xavier_uniform())
```
[the remark this answers is not in the document]
Comment (yooceii) [reply cleanrl/dqn_atari_jax.py]: Was checking dopamine's implementation when implementing the network. Removed.

**Comment by vwxyzjn** [comment]
This is awesome work. Thank you! @kinalmehta would you mind including the jitted action sampling function to #222? I think this will be the last thing before we merge. Since this is a non-breaking change for #222, we don't need to re-run the benchmark.

**Comment by vwxyzjn** [comment]
Closed in favor of https://github.com/vwxyzjn/cleanrl/pull/222

**Pull request closed without merging**